feat: integrate batch coalescer with repartition exec

jizezhang opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

- Part of https://github.com/apache/datafusion/issues/18782.

## Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

`RepartitionExec` has two cases: sort-preserving and non sort-preserving. This change integrates `LimitedBatchCoalescer` with the latter. For the former, it seems that `SortPreservingMergeStream` that builds on top of `PerPartitionStreams` has batching logic built in: https://github.com/apache/datafusion/blob/e4dcf0c85611ad0bd291f03a8e03fe56d773eb16/datafusion/physical-plan/src/sorts/merge.rs#L279-L289 hence I did not include in this change.

## What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

## Are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

Yes

## Are there any user-facing changes?

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

No

jizezhang commented on datafusion/core/tests/custom_sources_cases/provider_filter_pushdown.rs:
```diff
@@ -134,9 +134,19 @@ impl ExecutionPlan for CustomPlan {
         Ok(Box::pin(RecordBatchStreamAdapter::new(
             self.schema(),
-            futures::stream::iter(self.batches.clone().into_iter().map(Ok)),
+            futures::stream::iter(self.batches.clone().into_iter().map(move |batch| {
```
Please see https://github.com/apache/datafusion/issues/18782#issuecomment-3568216125 for my thoughts/reason on updating this test.

alamb commented:
🤖 Hi @Dandandan, thanks for the request (https://github.com/apache/datafusion/pull/19002#issuecomment-3595963851). [`scrape_comments.py`](https://github.com/alamb/datafusion-benchmarking/blob/main/scripts/scrape_comments.py) only supports whitelisted benchmarks: clickbench_1, clickbench_extended, clickbench_partitioned, clickbench_pushdown, tpch, tpch10, tpch_mem, tpch_mem10. Please choose one of these with `run benchmark <name>`.

Dandandan commented:
run benchmark tpch tpch_mem tpch10

martin-g commented on datafusion/physical-plan/src/repartition/mod.rs:
```diff
@@ -1540,7 +1550,46 @@ impl Stream for PerPartitionStream {
+                        }
+                        Poll::Ready(None) => {
+                            completed = true;
+                            coalescer.finish()?;
```
```suggestion
                            if let Err(e) = coalescer.finish() {
                                self.batch_coalescer = Some(coalescer);
                                return Err(e);
                            }
```
Otherwise in case of an error the `self.batch_coalescer` won't be restored.

jizezhang replied:
I think in this case, the loop would continue and then `break` on line  1561 or 1565. Then line 1588 is reached, which restores the `self.batch_coalescer`, but please let me know if that does not make sense.

jizezhang replied:
ah thanks for catching this and for the explanation, updated in revision.

alamb commented:
run benchmark clickbench_partitioned

jizezhang commented on datafusion/physical-plan/src/repartition/mod.rs:
```diff
@@ -1427,9 +1430,12 @@ struct PerPartitionStream {
 }
 
 impl PerPartitionStream {
+    #[allow(clippy::too_many_arguments)]
```
Thanks for pointing out, will update in a revision later.

jizezhang commented on datafusion/physical-plan/src/repartition/mod.rs:
```diff
@@ -1540,7 +1550,46 @@ impl Stream for PerPartitionStream {
+                            coalescer.finish()?;
+                        }
+                        Poll::Ready(Some(Ok(batch))) => {
+                            coalescer.push_batch(batch)?;
```
updated, thanks!

alamb commented on datafusion/physical-plan/src/repartition/mod.rs:
```diff
@@ -1427,9 +1430,12 @@ struct PerPartitionStream {
     /// Execution metrics
     baseline_metrics: BaselineMetrics,
+
+    batch_coalescer: Option<LimitedBatchCoalescer>,
```
I recommend adding some context here about when this could be null -- something like

```suggestion
    /// None for sort preserving variant (merge sort already does coalescing)
    batch_coalescer: Option<LimitedBatchCoalescer>,
```

alamb commented on datafusion/physical-plan/src/repartition/mod.rs:
```diff
@@ -1540,7 +1550,52 @@ impl Stream for PerPartitionStream {
+                    let inner_poll = self.poll_next_inner(cx);
+                    let _timer = cloned_time.timer();
+
+                    match inner_poll {
```
I found this logic slightly hard to follow and it is now in the main control flow loop and is deeply indented

One thing that could help here is to factor the limiting into its own function. So something like this

```rust
        mut poll;
        if let  Some(coalescer) = self.batch_coalescer.take() {
                poll = self.poll_next_and_colaesce(coalescer);
                self.batch_coalescer = Some(coalescer);
         } else {
                poll = self.poll_next_inner()
         }
        self.baseline_metrics.record_poll(poll)
```

And then the logic for handling the coalsecer could be put into `poll_next_and_colaesce`

jizezhang replied:
yes, makes sense, refactored in revision.

jizezhang commented:
> Thank you so much @jizezhang -- this is a really nice PR.
> 
> I had a few comments about the code, but I don't think any of them are required
> 
> The only thing I think is needed for this PR is a test showing that the coalescing is happening within the
> 
> I think you could perhaps follow the model of the existing tests in datafusion/physical-plan/src/repartition/mod.rs, and feed a bunch of small batches in (e.g. 5 rows) and then expect to get larger batches out the top
> 
> Again thank you for your work on this (and thank you @martin-g for the review)

Thanks a lot! I refactored the code based on your comment and also added unit test to test coalescing specifically.

jizezhang commented on datafusion/physical-plan/src/repartition/mod.rs:
```diff
@@ -1438,7 +1445,10 @@ impl PerPartitionStream {
         spill_stream: SendableRecordBatchStream,
         num_input_partitions: usize,
         baseline_metrics: BaselineMetrics,
+        batch_size: Option<usize>,
```
Hi @Dandandan , this was mentioned in the PR description, but it was because `RepartitionExec` has two variants: sort-preserving and non sort-preserving. For the former, as I understand, merge sort does batching https://github.com/apache/datafusion/blob/e4dcf0c85611ad0bd291f03a8e03fe56d773eb16/datafusion/physical-plan/src/sorts/merge.rs#L279-L293 Thus I made `batch_coalescer` and `batch_size` optional, e.g. only batch for non sort-preserving case. But please let me know if this makes sense to you. Thanks.

jizezhang commented on datafusion/physical-plan/src/repartition/mod.rs:
```diff
@@ -1531,6 +1542,43 @@ impl PerPartitionStream {
         }
     }
+
+    fn poll_next_and_coalesce(
```
Are you referring to maybe this method call https://github.com/apache/datafusion/blob/fc8824011bf5d4baccbfe51b3888ed5573ef3bfb/datafusion/physical-plan/src/repartition/mod.rs#L542 when pulling batches from input partitions? Do you mean that we could potentially combine it with coalescing? If yes, it was [discussed briefly](https://github.com/apache/datafusion/issues/18782#issuecomment-3563395564) on whether to coalesce batches in input partition stream or output partition stream. Current implementation coalesces in output stream, as it preserves existing behavior most. Since batches are sent over channels from input streams to output streams, I am not sure how we would combine. But I could have totally misunderstood you or it might actually be better to coalesce when pulling from input streams given the optimization. Please let me know what you think.

In the sort-preserving case, a `BatchBuilder` is used https://github.com/apache/datafusion/blob/fc8824011bf5d4baccbfe51b3888ed5573ef3bfb/datafusion/physical-plan/src/sorts/merge.rs#L44 which has methods such as `push_row` and `build_record_batch` https://github.com/apache/datafusion/blob/fc8824011bf5d4baccbfe51b3888ed5573ef3bfb/datafusion/physical-plan/src/sorts/builder.rs#L112-L125 which internally calls `interleave` from arrow. Would this also be something to be replaced/improved with the optimization you mentioned, or that is different?

Thanks!
